influxdb3_write/src/wal.rs

```diff
@@ -1,44 +1,642 @@
+    }
+
+    fn open_segment_reader(&self, segment_id: SegmentId) -> Result<Box<dyn WalSegmentReader>> {
+        let reader = WalSegmentReaderImpl::new(self.root.clone(), segment_id)?;
+        Ok(Box::new(reader))
+    }
+
+    fn segment_files(&self) -> Result<Vec<SegmentFile>> {
+        let dir = std::fs::read_dir(&self.root)?;
+
+        let mut segment_files = Vec::new();
+
+        for child in dir.into_iter() {
+            let child = child?;
+            let meta = child.metadata()?;
+            if meta.is_file() {
+                let path = child.path();
+                let file_name = path
+                    .file_stem()
+                    .expect("WAL segment files created by InfluxDB3 should have a file stem");
+                let file_name = file_name.to_str().expect("WAL segment files created by InfluxDB3 should have a file stem that is valid UTF-8");
+
+                if let Ok(segment_id) = segment_id_from_file_name(file_name) {
+                    segment_files.push(SegmentFile { segment_id, path });
+                } else {
+                    warn!(
+                        file_name,
+                        "WAL segment file has an invalid file stem, ignoring"
+                    );
+                }
+            }
+        }
+
+        segment_files.sort_by_key(|f| f.segment_id);
+
+        Ok(segment_files)
+    }
+
+    fn delete_wal_segment(&self, segment_id: SegmentId) -> Result<()> {
+        let path = build_segment_path(self.root.clone(), segment_id);
+        std::fs::remove_file(path)?;
+        Ok(())
     }
 }
 
-#[async_trait]
 impl Wal for WalImpl {
-    async fn open_segment_writer(
-        &self,
-        _segment_id: SegmentId,
-    ) -> crate::Result<Arc<dyn WalSegmentWriter>> {
-        todo!()
+    fn open_segment_writer(&self, segment_id: SegmentId) -> Result<Box<dyn WalSegmentWriter>> {
+        self.open_segment_writer(segment_id)
+    }
+
+    fn open_segment_reader(&self, segment_id: SegmentId) -> Result<Box<dyn WalSegmentReader>> {
+        self.open_segment_reader(segment_id)
     }
 
-    async fn open_segment_reader(
-        &self,
-        _segment_id: SegmentId,
-    ) -> crate::Result<Arc<dyn WalSegmentReader>> {
-        todo!()
+    fn segment_files(&self) -> Result<Vec<SegmentFile>> {
+        self.segment_files()
     }
 
-    async fn segment_files(&self) -> crate::Result<Vec<SegmentFile>> {
-        todo!()
+    fn delete_wal_segment(&self, _segment_id: SegmentId) -> Result<()> {
+        self.delete_wal_segment(_segment_id)
     }
+}
+
+#[derive(Debug)]
+pub struct WalSegmentWriterImpl {
+    segment_id: SegmentId,
+    f: File,
+    bytes_written: usize,
+    sequence_number: SequenceNumber,
+
+    buffer: Vec<u8>,
+}
+
+impl WalSegmentWriterImpl {
+    pub fn new_or_open(root: PathBuf, segment_id: SegmentId) -> Result<Self> {
+        let path = build_segment_path(root, segment_id);
+
+        let mut f = OpenOptions::new().write(true).create(true).open(&path)?;
```

Comment:
I think you can move this below the if statement and instead of `if f.len() > 0` just do `if path.exists()` this way you can avoid opening or creating the file twice.

Reply:
good call, fixed here: f32512754d11ec52e7e0a7627f204a7a94178221